adding new mysql shell backup engine

rvrangel opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

<!-- A few sentences describing the overall goals of the pull request's commits. -->
<!-- If this is a bug fix and you think the fix should be backported, please write so. -->

This is a PR that implements a new backup engine for use with MySQL Shell, which is mentioned in the feature request here: https://github.com/vitessio/vitess/issues/16294

It works a bit differently than the existing engines in vitess in which it only stores the metadata of how the backup was created (location + parameters used) and during the restore uses the location plus other parameters (mysql shell are different if you are doing a dump vs a restore, so we can't use exactly the same ones)



## Related Issue(s)

<!-- List related issues and pull requests. If this PR fixes an issue, please add it using Fixes #????  -->

Fixes #16294

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

shlomi-noach reviewed (commented):
Thank you for this submission! Some initial general thoughts, before a deeper code review. PErhaps these questions are more appropriate on https://github.com/vitessio/vitess/issues/16294 but I did not want to then split the discussion so let's keep it here.

I have not used MySQL Shell backups before. Some questions and notes:
- This PR adds dependencies on `mysqlsh` and `mysqlshell` binaries. This is just an observation, but points for consideration are:
  - Neither are included in a standard MySQL build. What are version dependencies between mysqlsh/mysqlshell and the MySQL server?
  - Neither are included in the MySQL docker images, to the best of my understanding. This means this backup method will not be available on kubernetes deployments via `vitess-operator`.

- Re: GTID not being available in the manifest file, this means we will not be able to run point in time recoveries with a mysqlshell-based full backup. Point in time recoveries require GTID information. As mentioned in https://github.com/vitessio/vitess/issues/16294#issue-2380771552, the mysqlshell method is the first and only (thus far) logical backup solution, so it's unfortunate that this solution will not support logical point in time recoveries.
  Is it not possible to read the `gtidExecuted` field from the `@.json` dump file immediately after the backup is complete, and update the manifest file? E.g. if the dump is into a directory, isn't that directory available for us to read?

rvrangel commented on go/vt/mysqlctl/backup.go:
```diff
@@ -500,3 +503,20 @@ func Restore(ctx context.Context, params RestoreParams) (*BackupManifest, error)
+		// This is usually run in a background goroutine, so there's no point
+		// returning an error. Just log it.
+		logger.Warningf("error scanning lines from %s: %v", prefix, err)
+	}
```
yeah, I can probably add it there 👍

rvrangel replied:
unit test added for this function

rvrangel commented:
These are good questions, thanks Shlomi!

- I wasn't sure, but checking the officila mysql docker images, it seems to be included actually:
    ```
    $ docker run -it mysql:8.0 mysqlsh --version
    Cannot set LC_ALL to locale en_US.UTF-8: No such file or directory
    mysqlsh   Ver 8.0.38 for Linux on x86_64 - for MySQL 8.0.38 (MySQL Community Server (GPL))
    ```
    In relation to the version dependency, my understanding is that MySQL Shell needs to be at least the same version of the MySQL Server, but it can be newer. we have successfully been using MySQL Shell 8.4 with Percona Server 8.0 releases.
    
    We don't use `vitess-operator` so for us it would mean we need to make sure required binaries are installed anyway (like `mysqld`, `xtrabackup`). But I imagine it being included in the official docker images means it will be less of an issue?
    
- Thats a good point I didn't realise. While it is possible to read the `@.json` file from a directory when it is completed (if we are writing to disk), it is less straight forward when we are storing the backups on an object store. Because `mysqlsh` doesn't work the same way (it doesn't provide you with a single stream that can be cleanly uplodad to whatever storage engine we are using), the thought was to bypass the storage engine in vitess (except for the `MANIFEST` which we still write using it) and just use this metadata to help the engine locate and restore the backup instead. If this was only saving to disk, it is much easier but also very limiting.
    
    If we were to do this, we would need to write code that would need to fetch the `@.json` from the supported object stores where there is a support overlap between `mysqlsh` and `vitess` (S3, Azure, GCP, etc) and some might be missing. Perhaps a better idea would be to include this backup engine without support for PITR in the beginning and file an upstream feature request that would print or save a copy of the executed GTID once the backup is done, which we could capture in an easier way (similar to the `xtrabackup` engine)?
    
     For additional context, as proposed on https://github.com/vitessio/vitess/pull/16428 and described in the use case, we plan to use this mostly to keep two backup types around for each shard, but always restoring by default from `xtrabackup` unless we require the logical backups for a specific reason.
     
     We also considered `mysqldump` which to be honest would fit the vitess backup engine workflow a lot better, but it was just too slow. [This benchmark from Percona](https://www.percona.com/blog/backup-restore-performance-conclusion-mysqldump-vs-mysql-shell-utilities-vs-mydumper-vs-mysqlpump-vs-xtrabackup/) also highlights the same thing, and for use backing up/restoring was so slow it didn't make sense.

frouioui reviewed (commented):
Hello, thank you for this contribution.

Since we have not fully finished the deletion of `mysqld` in the `vitess/lite` Docker Images, the `mysqlsh` binary will have to be included in the `vitess/lite` image regardless if it's included in the official MySQL Docker Images or not. Since we are letting people choose between using an official MySQL image or the `vitess/lite` image for their Docker/K8S deployment we must have the binary in both.

Regarding `vitess-operator`, a follow-up PR is needed on the vitess-operator to allow this new backup engine. In our CRDs we have an enumeration that restrict what backup engines are allowed, we just need to add a new entry in the enumeration. This can be done [here](https://github.com/planetscale/vitess-operator/blob/f754509b1ea682a88c632195b722360991dfd4f1/pkg/apis/planetscale/v2/vitesscluster_types.go#L319-L320).

FYI, I can handle the vitess-operator changes.

shlomi-noach commented:
> We also considered mysqldump which to be honest would fit the vitess backup engine workflow a lot better, but it was just too slow. 

Have you looked at `mysqlpump`? (Next gen `mysqldump`, included in standard builds).

shlomi-noach commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+		defer func() { // re-enable once we are done with the restore.
+			err := params.Mysqld.ExecuteSuperQueryList(ctx, []string{"ALTER INSTANCE ENABLE INNODB REDO_LOG"})
+			if err != nil {
+				params.Logger.Errorf("unable to re-enable REDO_LOG: %v", err)
```
Should we fail the restore process? I'm not sure if I have a good answer here.

rvrangel replied:
that's a good question. the original intention here is, since we were able to successfully disable Redo logs/double write buffer, it would be better to fail the restore than potentially put it in service with a potentially dangerous configuration without the user realising it. 

if the user wishes to run with redo log/double write disabled buffer they can avoid setting `--mysql_shell_speedup_restore` and handle it outside of vitess.

shlomi-noach replied:
Agreed, we should fail the restore when these flags are provided and if the above errors.

shlomi-noach replied:
Just verifying that this is still to be addressed.

rvrangel replied:
is there something you still wanted me to address? currently we will only attempt to disable/enable if the flags are passed, otherwise there is no change in behaviour and if we can't run the `ALTER INSTANCE` we will fail the backup

shlomi-noach replied:
> currently we will only attempt to disable/enable if the flags are passed, otherwise there is no change in behaviour and if we can't run the ALTER INSTANCE we will fail the backup

All right! Then there is nothing else to be addressed.

shlomi-noach commented:
> I wasn't sure, but checking the officila mysql docker images, it seems to be included actually:

Oh, that's nice! The reason I thought it wasn't included is that `mysqlsh`/`mysqlshell` is not included in the standard MySQL build.

> Thats a good point I didn't realise. While it is possible to read the @.json file from a directory when it is completed (if we are writing to disk), it is less straight forward when we are storing the backups on an object store.

I feel like it's OK to have some solution "with limitations". We should strive to support as much functionality as possible though. So IMHO we should strive to include the GTID when the backup goes into a directory. This should be possible to do, which then means the backup should fail if for some reason we can't fetch the GTID or validate it (correct GTID form). i.e. return `BackupUnusable` if unable to fetch and validate the GTID entry.

I'd like @deepthi to weigh in her opinion. 

Assuming we do decide to move forward, then I'd next expect a CI/endtoend test please, as follows:

- Take a look at https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/go/test/endtoend/backup/pitr_xtrabackup/backup_pitr_xtrabackup_test.go and https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/go/test/endtoend/backup/pitr/backup_pitr_test.go. You will create a similarly third path&file that uses a (new)  `SetupType:      backup.MySQLShellBackup`.
- Similarly to https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/.github/workflows/cluster_endtoend_backup_pitr.yml, there needs to be a test for `backup_pitr_mysqlshell`
- Also configured in [test/config.json](https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/test/config.json#L94-L111).

When these are all added, a new CI job will run to test `mysqlshell`-based backup, restores, and point-in-time recoveries. These can (and should) use the directory-based backup configuration, one which does make the GTID available.

If this test passes, then you will have validated the full cycle of backup and resotre, as well as correctness of the captured GTID.

S3, Azure, GCP can be left without GTID support for now.

We'd need a documentation PR that clearly indicates the limitations of this method.

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+	// flags passed to the mysql shell utility, used both on dump/restore
+	mysqlShellFlags = "--defaults-file=/dev/null --js -h localhost"
+	// flags passed to the Dump command, as a JSON string
+	mysqlShellDumpFlags = `{"threads": 2}`
```
I would say because MySQL Shell backups can be a bit more intensive - we are requesting a bunch of data off MySQL which needs to be fetched, parsed and compressed - and in our particular use case we were taking backups online so we didn't wan't it to cause that much disruption. It is also part of the reason why I made sure  `ShouldDrainForBackup()` was configurable in case it is more suitable for the use case.

I am fine with changing the default to `runtime.NumCPU()` though since it is configurable and leave this up to the user to decide based on their environment requirements, although I am also conscious that it might cause some issues in Kube where it will show the number of CPUs of the underlying node despite the pod being possibly limited on how much CPU it can use.

shlomi-noach replied:
Seems like we already have a `--concurrency` flag for the `Backup` command. Should we reuse that flag?

rvrangel replied:
so looking this up again today, the default in MySQL Shell if not passed is `4`. what I have been considering is that if we do something like:
```go
mysqlShellDumpFlags = fmt.Sprintf("{\"threads\": %d}", runtime.NumCPU())
```

it might give the user the wrong impression that this is the default for MySQL Shell and that if they use the flag to change other settings and they don't add `thread` to it, they will use all CPUs by default. 

so perhaps we should just set it to `{"threads": 4}` instead?

shlomi-noach replied:
> it might give the user the wrong impression that this is the default for MySQL Shell and that if they use the flag to change other settings and they don't add thread to it, they will use all CPUs by default.

This is a documentation concern. We can't expect any user expectation. I can say that I would not assume as much.

rvrangel replied:
still, we can try to make it more difficult for users to do the wrong thing. perhaps we can remove `threads` from the default value and leave it up to the user to modify it if the desire so?

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+	fs.StringVar(&mysqlShellFlags, "mysql_shell_flags", mysqlShellFlags, "execution flags to pass to mysqlsh binary to be used during dump/load")
+	fs.StringVar(&mysqlShellDumpFlags, "mysql_shell_dump_flags", mysqlShellDumpFlags, "flags to pass to mysql shell dump utility. This should be a JSON string and will be saved in the MANIFEST")
+	fs.StringVar(&mysqlShellLoadFlags, "mysql_shell_load_flags", mysqlShellLoadFlags, "flags to pass to mysql shell load utility. This should be a JSON string")
+	fs.BoolVar(&mysqlShellBackupShouldDrain, "mysql_shell_should_drain", mysqlShellBackupShouldDrain, "decide if we should drain while taking a backup or continue to serving traffic")
```
yeah, exactly. in fact I have been meaning to propose this to be modifiable for the `xtrabackup` engine as well, so a tablet won't be service traffic when it is taking a backup

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+	fs.StringVar(&mysqlShellDumpFlags, "mysql_shell_dump_flags", mysqlShellDumpFlags, "flags to pass to mysql shell dump utility. This should be a JSON string and will be saved in the MANIFEST")
+	fs.StringVar(&mysqlShellLoadFlags, "mysql_shell_load_flags", mysqlShellLoadFlags, "flags to pass to mysql shell load utility. This should be a JSON string")
+	fs.BoolVar(&mysqlShellBackupShouldDrain, "mysql_shell_should_drain", mysqlShellBackupShouldDrain, "decide if we should drain while taking a backup or continue to serving traffic")
+	fs.BoolVar(&mysqlShellSpeedUpRestore, "mysql_shell_speedup_restore", mysqlShellSpeedUpRestore, "speed up restore by disabling redo logging and double write buffer during the restore process")
```
unless you need the redo log/double write buffer to be disable once the instance has completed the restore, there shouldn't be much risk in enabling this. for some setups there might be an interest in disabling this (I can see as a possible case, somebody running on `zfs` and wanting to keep the double write buffer disabled), so I didn't want to force it if the user would rather use those

shlomi-noach replied:
OK, so for the duration of the restore this is fine, because who cares, the server is not serving, it's just being restored, right? But then, at the end of the process, as we mentioned elsewhere, these settings must return to "persistent" values, or else the backup should fail.

shlomi-noach replied:
> One thing that I realise now is that this will likely fail if the user is running <8.0.21. Is this something we should validate in case this flag is set or do we already have a minimum version for MySQL with Vitess?

We should validate that based on the actual MySQL connection. This is an already recognized "capability" in https://github.com/vitessio/vitess/blob/ae7214d9bcd4db23c7aeeb566757951501aba158/go/mysql/capabilities/capability.go#L46

Here is an example usage:

https://github.com/vitessio/vitess/blob/ae7214d9bcd4db23c7aeeb566757951501aba158/go/vt/vttablet/onlineddl/executor.go#L851-L852

We should fail the operation if the flags are set and the MySQL server versions is "incapable of" the functionality. Ideally, as close as possible to the flag parsing, but also just fine if just as part of the restore process, but do try to move that up as much as possible, and before doing actual substantial resotre work (e.g. before destorying the existing data...)

rvrangel replied:
this has been added on 6e59215d34e9d526e937047409ecea5a263e02c3

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+	}
+
+	start := time.Now().UTC()
+	location := path.Join(mysqlShellBackupLocation, params.Keyspace, params.Shard, start.Format("2006-01-02_15-04-05"))
```
ah, simply because we were already using another alternative to taking logical backups outside of vitess and this follows the same format.

but to be fair, vitess already doesn't use a standard format either. backups are created using:
https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/go/vt/mysqlctl/backup.go#L59-L60

I can switch to using that format if it makes more sense?

deepthi replied:
Yes, let us use this format, and keep it consistent within the backups code.

shlomi-noach replied:
And I agree the format used by vitess is also non standard (and weird looking).

rvrangel replied:
this was changed to use the existing format!

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+		args = append(args, strings.Fields(mysqlShellFlags)...)
+	}
+
+	args = append(args, "-e", fmt.Sprintf("util.dumpSchemas([\"vt_%s\"], %q, %s)",
```
you mean escape like in MySQL as `` `vt_keyspace` ``? I am not sure MySQL Shell supports or expects it, I will verify that

deepthi replied:
You can't assume that the database name is `vt_keyspace`. It is determined by the `--init_db_name_override` flag.

deepthi replied:
This brings up another point. Typically when we do the regular kind of backup we also backup the sidecar db, which defaults to `_vt`. Does the mysqlshell backup specifically backup only the actual keyspace/database? What are the implications of this?

rvrangel replied:
oh, it is possible to backup multiple db/schemas, we just need to specify them. is this the right place to get the name of the sidecar db?
https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/go/constants/sidecar/name.go#L23-L25

also, I am curious what are the consequences if `_vt` is not in place, we tried some of these restores without any issues but maybe we are losing some metadata?

rvrangel replied:
moved the code to call `sidecar.GetName()`

deepthi replied:
Re the sidecar db, you may not see any bad effects unless and until you try to promote the new replica, or unless you are in the midst of a reshard / online DDL etc. Also schema tracking. If any of these features are active, the replication stream could eventually break after the replica is up.

deepthi replied:
@rvrangel it seems like `vt_` + keyspace name is still hardcoded. That needs to be fixed. You will need to somehow look that up and thread it through.

shlomi-noach replied:
@rvrangel on that note, something I only realized when discussing your PR with the team, are you taking backups of explicit databases? Is there a way in `mysqlshell` to abckup all databases? I'd assume the `mysql` database needs backing up?

rvrangel replied:
that's a good question. we can backup all databases (that's the difference between `dumpInstance()` vs `dumpSchemas()` if I remember correctly), I guess for our setup we create mysql grants every time we provision the tablets in a different way, is that why you are thinking of backing up the `mysql` DB too?

we could just add it to the list if other databases are supposed to be ephemeral from the Vitess point of view, otherwise would backing up all of them be preferable?

shlomi-noach replied:
Vitess-wise we should backup all databases. When vitess restores MySQL from backup, it makes no assumptions on existing data, nor does it populate any specific data prior/after backup. It expects the backup to just work and restore MySQL correctly.

rvrangel replied:
okay, I have updated it to use `dumpInstance()` instead, which should include all the databases now.

rvrangel replied:
@shlomi-noach one interesting side effect of this I was thinking when [discussing a related issue](https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2324490828) on this other PR, is that we now dump all databases, but we only drop the ones managed by vitess on restore and that could causes during restore since we only drop the ones managed by vitess:
https://github.com/vitessio/vitess/pull/16295/files#diff-7445e78fc6fe2a32a3e8b6a01f6159a07e3b79d29f83eb81b22f3601a1e91436R275-R277

what do you believe is the best course of action for vitess to take here? leave it as is (so the user will need to clear MySQL before any restore that already has data on disk if they are using other databases) or then make sure we are dropping all databases before the restore takes place?

shlomi-noach replied:
Right. So this is something completely new to the vitess backup/restore process, and I did not catch that earlier.

With both existing backup engines (`builtin` and `xtrabackup`) Vitess assumes the entire schema, including `mysql.*` tables, is restored from backup in one shot. This also aligns with GTID positioning as `gtid_executed` is just restored as part of the entire table set in `mysql.gtid_executed`.

So what I'm missing right now is how `mysqlshell` backs up and restores. From a [quick read](https://www.percona.com/blog/backup-and-restore-using-mysql-shell/) I gather `mysqlshell` doesn't actually backup` the `mysql` database? Or can it be forced to do so?

And, upon restore, am I correct to understand there needs to be a database up and running, and with some appropriate user privileges, before issuing a restore? What are the minimal requirements?

The most vitess-compliant method would be to issue a complete dump and complete restore. Is there a strong argument against doing so, and for only backing up a subset of databases? I don't see how that would play nicely with GTID, but if there's a known solution, please let me know.

rvrangel replied:
includeSchemas, `dumpInstance()` [doesn't support backing up these even if using](https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-dump-instance-schema.html#mysql-shell-utilities-dump-opt-filtering) `includeSchemas`:

>  includeSchemas: array of strings
>
>    (Instance dump utility only) Include only the named schemas in the dump. This option is available from MySQL Shell 8.0.28. You cannot include the information_schema, mysql, ndbinfo, performance_schema, or sys schemas by naming them on this option. If you want to dump one or more of these schemas, you could do this using the schema dump utility util.dumpSchemas()

I think this is less of an issue for us as we deal with MySQL grants outside of Vitess, and not every table will be replicated, but @ejortegau and I were brainstorming on it and one possible solution if you expect grants to be handled on the mysql database is to use `loadUsers` option in MySQL Shell. MySQL Shell already includes users by default on the backups with `dumpInstance()`, so we could just restore the users as they were when the backup is taken, but that requires us deleting all the users currently on MySQL (except the current user) otherwise it will error out if they already exist, [see the docs](https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-load-dump.html#mysql-shell-utilities-load-dump-opt-filtering):

> loadUsers: [ true | false ]
>
>    Import (true) or do not import (false) users and their roles and grants into the target MySQL instance. The default is false, so users are not imported by default. Statements for the current user are skipped. From MySQL Shell 8.0.22, if a user already exists in the target MySQL instance, an error is returned and the user's grants from the dump files are not applied. From MySQL Shell 8.0.22, you can use the excludeUsers or includeUsers option in the dump loading utility to specify user accounts to be excluded or included in the import. 

Let me know if you think that would ease the concerns with the GTID breaking under this. as mentioned above, we manage grants outside of vitess and we avoid replication with `sql_log_bin=0` when making changes there, so we haven't seen an issue with it so far

frouioui replied:
I am answering to https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2324490828 here so everything is in the same place.

> could you provide a bit more of detail of how it is re-created? when you say the data is restored from disk, do you mean you restore previous snapshots of the data (so tables and schema would already exist on disk) while starting the vttablet with --restore-from-backup?

That's correct.

The cluster is running, we then teardown the vttablets, keeping the vtdataroot and all associated data, which is what happen by default when you remove the tablets on kubernetes. We then bring the tablets back up again, and they start with the vtdataroot already filled, meaning the data will be loaded from what already exists there. Moreover, because we defined a backup configuration on our K8S yaml file, we will start vttablet with the `--restore_from_backup` flag. Which will try to load any existing backup after the tablets are up and running.


> if so, the main issue [is in here](https://github.com/vitessio/vitess/pull/16295/files#diff-7445e78fc6fe2a32a3e8b6a01f6159a07e3b79d29f83eb81b22f3601a1e91436R251-R283), we only drop those databases for receiving a RestoreFromBackup() API call, not when starting with --restore-from-backup.

Would it be possible to drop those databases always when using the mysqlsh engine?

> from the code comments, it seems if you are starting a tablet with existing data, [it needs to be wiped out before you start the ](https://github.com/slackhq/vitess/blob/main/go/vt/mysqlctl/backupengine.go?rgh-link-date=2024-09-02T11%3A21%3A35Z#L113-L116)vttablet or you need to start the restore via an API call instead of relying on --restore-from-backup.

I don't think wiping out the data manually in vtop would be a good idea, we could explore the idea of not starting vttablet with the restore from backup flag and instead do an API call when the tablet is healthy, but I would like to know if we could simplify it by simply dropping the databases always when mysqlsh engine is used.

rvrangel replied:
it would be possible, but it seems to me it would be an exception to handling `DeleteBeforeRestore`, where its value is false but we go and delete files anyway.

looking at the other backup engines, they delete current files without caring for the value of `DeleteBeforeRestore`. I can't find where this is actually used rather than just passed down all the way to the backup engine, but ignored on the other two. is there something I am missing? if we have to delete the databases anyway, I would prefer to just get rid of `DeleteBeforeRestore` if not being used, rather than having the backup engines ignore the intended value.

shlomi-noach replied:
This looks to be the only place where the value of `DeleteBeforeRestore ` is checked:

https://github.com/vitessio/vitess/blob/44e96c2291a2b7b4fddbd7162d2214ba58661810/go/vt/mysqlctl/backup.go#L311-L314

And it looks to me like it is senseless. I think the variable can be removed.

rvrangel replied:
@shlomi-noach would it make sense to do it in a separate PR though, so it is easier to revert if needed? I would hate for us to have to revert this PR once merged in case it causes any side effects

rvrangel replied:
`DeleteBeforeRestore` aside, would you all be comfortable with the idea adding a flag that lets makes us run `DELETE USER` on all users except the current one and than letting the user define `loadUsers` on the mysql shell restore flags on their own so these are reloaded as appropriate if they rely on replication for propagating the grants?

shlomi-noach replied:
> I would prefer to just get rid of DeleteBeforeRestore if not being used, rather than having the backup engines ignore the intended value.

I interpreted that in the sense that you prefer to remove the variable in this PR. It's fine if we did that on another PR.

frouioui replied:
> would it make sense to do it in a separate PR though
Only thing to take into account is that if we do it in a separate PR, after merging this current PR the new engine won't work on vitess-operator (the restore part only). I don't mind waiting for the second PR to be ready/merged in order to merge the vitess-operator one.

rvrangel replied:
okay, lets do it in this PR then to avoid the admin work of having it separate and putting up notices, I will push a change later today

rvrangel commented:
@shlomi-noach yeah, we looked into `mysqlpump`, but it has been deprecated (the page you linked also has a notice) and it is still slower than MySQL Shell. but since it is likely going to be removed in a future MySQL version, we though it would be better not to introduce a new feature using it :) 

I think the proposal to have the GTID read when backing up to a directory while not when using an object store is fair, let me look into it and make the necessary changes. If all looks good I will proceed with working on the CI/endtoend tests, I just wanted to get some initial feedback before doing so. Also curious what @deepthi thinks about this approach.

> Edit: since mysqlshell does not come bundled in the mysql distribution, we'd need to further download/install mysqlshell in the GitHub workflow file.
Is that something that needs to happen as part of this PR or something separate?

rvrangel commented:
> Since we have not fully finished the deletion of mysqld in the vitess/lite Docker Images, the mysqlsh binary will have to be included in the vitess/lite image regardless if it's included in the official MySQL Docker Images or not. Since we are letting people choose between using an official MySQL image or the vitess/lite image for their Docker/K8S deployment we must have the binary in both.

@frouioui  We don't use the docker images, but I am not sure I understand, if `mysqld` is being removed does that means users are expect to use a different image for running `mysqld`?

regarding the change in the `vitess-operator`, I opened a separate PR for it: https://github.com/planetscale/vitess-operator/pull/586, is that all that needs to be changed?

shlomi-noach commented:
>  I just wanted to get some initial feedback before doing so. Also curious what @deepthi thinks about this approach.

@rvrangel absolutely! Let's wait for more input.

>> Edit: since mysqlshell does not come bundled in the mysql distribution, we'd need to further download/install mysqlshell in the GitHub workflow file.

> Is that something that needs to happen as part of this PR or something separate?

In the scope of this PR please. I was merely pointing out that our standard workflow won't install mysql-shell. Oh, and there's a greater discussion about how our workflows are generated, see `test/ci_workflow_gen.go`. I can help you with that part if you need.

frouioui commented:
> We don't use the docker images, but I am not sure I understand, if `mysqld` is being removed does that means users are expect to use a different image for running `mysqld`?

@rvrangel, historically we have always shipped `mysqld` and all required binaries in our Docker Images (`vitess/lite`). My point is that we need to make sure `mysqlsh` is available in the `vitess/lite` Docker Image so that people can use all the Vitess features on Docker/K8S. Currently, `mysqlsh` is not in our `vitess/lite` Docker Image:
```
$> docker run -it --user=vitess vitess/lite:latest bash
vitess@e92ffa0b2f14:/$ mysqlsh
bash: mysqlsh: command not found
```

I am working on a fix for it, I will push directly to this branch.


> regarding the change in the `vitess-operator`, I opened a separate PR for it: [planetscale/vitess-operator#586](https://github.com/planetscale/vitess-operator/pull/586), is that all that needs to be changed?

Thank you for opening this! 🙇🏻 I will take a look and comment on that PR directly.

deepthi commented:
@shlomi-noach and @frouioui have pretty much covered all of the main concerns. My opinion is that in principle this is a good feature addition.
The only concern I have left is whether this feature works only with the file backup storage type. I'm basing this off shlomi's comment
> S3, Azure, GCP can be left without GTID support for now.

Can you explain what will and will not work with S3, GCP and Azure?

frouioui reviewed (commented):
I think it might be worth writing something about this new backup engine in the release notes for v21.0.0: `./changelog/21.0/21.0.0/summary.md`.

frouioui commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+package mysqlctl
+
```
This file and `go/vt/mysqlctl/mysqlshellbackupengine_test.go` need a licence header.

deepthi commented on go/flags/endtoend/vtbackup.txt:
```diff
@@ -178,6 +178,12 @@ Flags:
       --mysql-shutdown-timeout duration                             how long to wait for mysqld shutdown (default 5m0s)
       --mysql_port int                                              mysql port (default 3306)
       --mysql_server_version string                                 MySQL server version to advertise. (default "8.0.30-Vitess")
+      --mysql_shell_backup_location string                          location where the backup will be stored
+      --mysql_shell_dump_flags string                               flags to pass to mysql shell dump utility. This should be a JSON string and will be saved in the MANIFEST (default "{\"threads\": 2}")
+      --mysql_shell_flags string                                    execution flags to pass to mysqlsh binary to be used during dump/load (default "--defaults-file=/dev/null --js -h localhost")
+      --mysql_shell_load_flags string                               flags to pass to mysql shell load utility. This should be a JSON string (default "{\"threads\": 4, \"updateGtidSet\": \"replace\", \"skipBinlog\": true, \"progressFile\": \"\"}")
+      --mysql_shell_should_drain                                    decide if we should drain while taking a backup or continue to serving traffic
+      --mysql_shell_speedup_restore                                 speed up restore by disabling redo logging and double write buffer during the restore process
```
New flags need to use dashes as separators.

shlomi-noach commented:
>> S3, Azure, GCP can be left without GTID support for now.

> Can you explain what will and will not work with S3, GCP and Azure?

@deepthi I will meanwhile explain what my understanding is: that in S3, Azure, and GCP, we won't have GTID information in the backup manifest, which means we won't be able to use such backups for point in time recoveries.

frouioui reviewed (commented):
Copying this here so other reviewers are aware of it, we need to update the `operator.yaml` file in the Vitess repo with the changes brought by the corresponding PR on the vitess-operator PR: https://github.com/planetscale/vitess-operator/pull/586

frouioui commented on docker/lite/Dockerfile:
```diff
@@ -42,7 +42,7 @@ RUN /vt/dist/install_dependencies.sh mysql80
 # Set up Vitess user and directory tree.
 RUN groupadd -r vitess && useradd -r -g vitess vitess
-RUN mkdir -p /vt/vtdataroot && chown -R vitess:vitess /vt
+RUN mkdir -p /vt/vtdataroot /home/vitess && chown -R vitess:vitess /vt /home/vitess
```
Sorry for coming back to this only now, but I think we should get rid of the warning generated by `mysqlsh` when it cannot find the `en_US.UTF-8` locale: `Cannot set LC_ALL to locale en_US.UTF-8: No such file or directory`.

I came up with a fix for the warning, which installs the required locale but does not set it as a default locale, which should not change the behavior of the image and Vitess. Here is the code snippet, which should be added to this file:

```Dockerfile
# Install locale required for mysqlsh
RUN apt-get update && apt-get install -y locales \
    && echo "en_US.UTF-8 UTF-8" > /etc/locale.gen \
    && locale-gen en_US.UTF-8
```

I tested locally and `mysqlsh` is no longer complaining about the locale. By doing this the image size grows by approximately 2% as seen in the logs below.

```
# Before the fix
$> docker images ls
...
vitess/lite-test              latest             0c4d123a8f79   24 hours ago    2.09GB
...
$> docker run -it --user="vitess" 0c4d123a8f79 bash
vitess@8e6183dc9343:/$ mysqlsh --version 
Cannot set LC_ALL to locale en_US.UTF-8: No such file or directory
mysqlsh   Ver 8.0.37 for Linux on x86_64 - for MySQL 8.0.37 (MySQL Community Server (GPL))
vitess@8e6183dc9343:/$

# After the fix
$> docker build -t vitess/lite-test:latest -f docker/lite/Dockerfile .
$> docker images ls
...
vitess/lite-test              latest             82391c5265a9   2 minutes ago   2.13GB
...
$> docker run -it --user="vitess" docker.io/vitess/lite-test:latest bash 
vitess@05bafcc52051:/$ mysqlsh --version 
mysqlsh   Ver 8.0.37 for Linux on x86_64 - for MySQL 8.0.37 (MySQL Community Server (GPL))
vitess@05bafcc52051:/$
```

rvrangel replied:
thanks for providing the previous patch @frouioui, it is much appreciated. I am adding this as well as part of the next commit

frouioui replied:
This looks good to me now. Closing this comment. Thank you!

rvrangel commented:
@shlomi-noach sorry I haven't made too much progress on this today, but wanted to bring another idea for something we discussed today in relation to the GTIDs.

what do you think if instead of capturing the executed GTID sets from the `@.json` file, we were to stop replication on the tablet, capture the executed GTID sets directly from MySQL and then starting the backup process? once MySQL Shell has acquired the global read lock, it should have the same GTID set we had and then we can start the replication threads again. Any thoughts on this approach?

shlomi-noach commented:
> sorry I haven't made too much progress on this today

We're here for the long run. Take your time.

> what do you think if instead of capturing the executed GTID sets from the @.json file, we were to stop replication on the tablet

That's a good idea, with questions. Does that mean backups need to necessarily run on replicas? Or will `mysqlshell` also be able to run on primaries? Meaning, is this a new limitation we will now impose on `mysqlshell` backups, or is that an already existing limitation?
I'm cheating a bit in this question, because we normally run backups from a `BACKUP` tablet type, which is always a replica.

When taking a backup from a replica, if you need to stop replication, which is fine, you first need to take the replica out of serving mode, to return it back. Usually the tablet type is `BACKUP` anyway, which means it's not serving, but it _could_. Let's do  the comparison with `builtin` backup engine, that goes farther to actually shut down the mysql server. See what happens there in terms of draining the tablet / taking it out of and back into service?

rvrangel commented:
fair enough, we don't use `vtbackup` so we don't actually provision an entirely new tablet which will be of the `BACKUP` type. saying that, there are some differences in between the engines. the `builtin` backup engine will automatically transition the host to `BACKUP` because ` ShouldDrainForBackup()` returns `true`, which the `xtrabackup` engine doesn't.

So I guess our options are:
#### collect the GTIDs from the output directory
- can be used on any tablet type
- can be taken online, similar to the `xtrabackup` engine
- won't support PITR for object stores, only when backing up to a directory

#### stop replication and collect GTIDs from MySQL
- backup can't be taken from a primary
- replication needs to be stopped (either briefly at the beginning or for the entire backup) to ensure the GTID sets are consistent
- will support PITR regardless of where it is stored.

second approach sounds like a good compromise, but first I wanted to ask a clarifying question. while the GTID sets are not in the `MANIFEST`, we actually pass them back to Vitess when it is doing a restore (that's because we have time to fetch them from MySQL between MySQL Shell has finished the restore and we hand it over to Vitess). That means the `ExecuteRestore()` will return the `MANIFEST` with the correct GTIDs then. Does the Vitess PITR process expect to be able to read the GTIDs from the `MANIFEST` before it attempts to restore any backups, or does it happen later in the process?

shlomi-noach commented:
> Does the Vitess PITR process expect to be able to read the GTIDs from the MANIFEST before it attempts to restore any backups

Yes, it does, as part of computing the potential recovery path (one full restore + zero or more incremental restores).

rvrangel commented:
I hit an issue that we can't start/stop replication when there is a backup lock in place in MySQL (created by MySQL Shell), so we wouldn't be able to use that approach.

a different solution I have been testing is to acquire a global read lock (`FLUSH TABLES WITH READ LOCK`) and the releasing it as soon `mysqlsh` releases its own, this way we only lock writes for a couple of seconds and we get the consistent GTID sets - it also works on all tablet types including primaries. I am making adjustments to make sure the `endtoend` tests cover everything.

rvrangel commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,340 @@
+	// location to store the mysql shell backup
+	mysqlShellBackupLocation = ""
+	// flags passed to the mysql shell utility, used both on dump/restore
+	mysqlShellFlags = "--defaults-file=/dev/null --js -h localhost"
```
this is left for the operator to decide, depending on their environment on how to connect to `mysqld`. I am happy to change the defaults to something else (e.g. unix socket path)

mattlord replied:
I think it's probably best NOT to specify a default here (and `-h localhost` is already the compiled default I believe). I'm assuming these are easily overridden no matter what we do here though? 

If we do add a default here, we might want to use the default socket file path in the operator: https://github.com/planetscale/vitess-operator/blob/58a8d79bb44e4238a274598817e910c57be60950/pkg/operator/vttablet/constants.go#L71

That being: `/vt/socket/mysql.sock`

frouioui replied:
I can indeed confirm that this is not working: https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2313300307

rvrangel replied:
this is the default value and should be adjusted to match the running environment (for example, if running on kube, provide the host/port or unix socket file), similar to what is done in the end to end tests. would not providing the `-h` flag be better in this case? AFAIK mysql shell will still use `localhost` as the default value

frouioui replied:
I am fine keeping this default value on vitess, however on the vitess-operator we should set another value that overrides the default of `--mysql-shell-flags`. That way when running with vitess-operator `localhost` will be replaced by the socket.

frouioui replied:
Closing this as it has been resolved on the vtop side.

rvrangel commented:
@shlomi-noach I have added the `endtoend` tests and fixed the issues with the other tests, I would appreciate if you could take a look when you have the time. I changed the approach as described in my last comment, which will support different tablet types (including `PRIMARY`) and will cause the least amount of disruption to the tablet, while being able to consistently save the GTID sets and work with PITR

shlomi-noach commented:
@rvrangel awesome, thank you! I'll do another full review by early next week (already done for the week).

deepthi commented:
> I hit an issue that we can't start/stop replication when there is a backup lock in place in MySQL (created by MySQL Shell), so we wouldn't be able to use that approach.

The other thing you would have ended up hitting is vtorc restarting replication that was stopped by this process. So I'm glad you found an alternate solution.

shlomi-noach commented on go/test/endtoend/backup/pitr_mysqlshell/backup_pitr_mysqlshell_test.go:
```diff
@@ -0,0 +1,61 @@
+func TestIncrementalBackupAndRestoreToTimestamp(t *testing.T) {
+	tcase := &backup.PITRTestCase{
+		Name:      "MySQLShell",
+		SetupType: backup.XtraBackup,
```
Should this be `backup.MySQLShell`?
```suggestion
		SetupType: backup.MySQLShell,
```

shlomi-noach commented on go/test/endtoend/backup/vtctlbackup/backup_utils.go:
```diff
@@ -54,19 +54,21 @@ const (
+	localCluster     *cluster.LocalProcessCluster
+	newInitDBFile    string
+	useXtrabackup    bool
+	currentSetupType int
```
`currentSetupType` looks good. I see it replaces `useXtrabackup` in a few places -- should it completely replace it, such that we can remove `useXtrabackup` altogether?

rvrangel replied:
I was a bit unsure about how the removal of the last pieces of `useXtrabackup` would play out, but I did on the last commit and it seems tests are passing locally, but if you could take a better look I would appreciate.

shlomi-noach commented on go/vt/mysqlctl/backup_test.go:
```diff
@@ -714,3 +713,26 @@ func TestShouldRestore(t *testing.T) {
 	assert.NoError(t, err)
 }
+
+func Test_scanLinesToLogger(t *testing.T) {
```
Use CamelCase for function names, even `Test` ones.
```suggestion
func TestScanLinesToLogger(t *testing.T) {
```

shlomi-noach reviewed (commented):
Thank you for adding `endtoend` tests! To get those to actually run, you'll need the following:

1. Add `pitr_mysqlshell` in `test/ci_workflow_gen.go` like so: https://github.com/vitessio/vitess/blob/44e48cad10ad8e73733bd86ea917f69218059c94/test/ci_workflow_gen.go#L76-L77
2. Run `make generate_ci_workflows`
3. Find the new `.github/workflows/...` file, add and commit.
4. The new CI is expected to run on the next push.

shlomi-noach commented:
- [ ] When we get to adding release notes, we should mark this new feature as "experimental"

shlomi-noach commented:
Could you please rename `pitr_mysqlshell` to `backup_pitr_mysqlshell` so that it aligns with `backup_pitr_extrabackup`?

rvrangel commented:
yeah, I actually fixed it to be `backup_pitr_mysqlshell` but I didn't re-generate them, it should be correct now

shlomi-noach commented on go/test/endtoend/backup/vtctlbackup/backup_utils.go:
```diff
@@ -1420,8 +1445,11 @@ func verifyTabletRestoreStats(t *testing.T, vars map[string]any) {
 	// Currently only the builtin backup engine emits operation counts.
-	if !useXtrabackup {
+	case XtraBackup, MySQLShell:
+	default:
```
here, too, I'd use `case XtraBackup, MySQLShell:`

shlomi-noach commented on go/test/endtoend/backup/vtctlbackup/backup_utils.go:
```diff
@@ -1432,8 +1460,11 @@ func verifyTabletRestoreStats(t *testing.T, vars map[string]any) {
 	// Currently only the builtin backup engine emits timings.
-	if !useXtrabackup {
+	case XtraBackup, MySQLShell:
+	default:
```
again, `case XtraBackup, MySQLShell:`

rvrangel commented on go/test/endtoend/backup/vtctlbackup/backup_utils.go:
```diff
@@ -1373,8 +1392,11 @@ func verifyTabletBackupStats(t *testing.T, vars map[string]any) {
 	// Currently only the builtin backup engine emits timings.
-	if !useXtrabackup {
+	case XtraBackup, MySQLShell:
+	default:
```
this is a good idea, just to be clear as I wasn't sure by your comments below, after changing the `default` to `case BuiltinBackup:`, do you prefer I remove the empty `case XtraBackup, MySQLShell:`  or leave it there to be more explicit?

rvrangel replied:
cool, I have removed it now 👍

shlomi-noach commented:
I see the new `backup_pitr_mysqlshell` passes CI -- this is wonderful news! I'll look into the logs to get more details.

rvrangel commented:
it seems the static checks are failing now, complaining about different version of Go in the CI templates, but I don't remember making any change that would cause that, would know what the problem could be? running `tools/check_go_versions.sh` doesn't produce any errors

shlomi-noach commented:
> it seems the static checks are failing now, complaining about different version of Go in the CI templates, but I don't remember making any change that would cause that

Quickest advice: merge the latest from `main`? Hopefully that will make the problem go away. Otherwise we'll have to look into it 😆

shlomi-noach commented:
> merge the latest from main

And make sure to then `make generate_ci_workflows` again.

rvrangel commented:
yeah, that fixed it, all tests are passing now 🎉

frouioui commented:
Gentle reminder @rvrangel in case this got lost among all the other comments: https://github.com/vitessio/vitess/pull/16295#pullrequestreview-2194441789 is still pending.

Let me know if I can help in any way with the vitess-operator stuff. I am happy to help 💪🏻

deepthi commented:
@rvrangel In addition to what @frouioui pointed out, there are two other things that are needed before we can merge
- an update to the release notes summary file with a brief note on this functionality
- a website docs PR that updates all relevant backup/restore pages in the v21 (Development) docs.

frouioui commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,510 @@
+
+func (be *MySQLShellBackupEngine) backupPreCheck(location string) error {
+	if mysqlShellBackupLocation == "" {
+		return fmt.Errorf("%w: no backup location set via --mysql_shell_location", MySQLShellPreCheckError)
```
This error needs to use the new flag format (`--mysql-shell-backup-location`)

```suggestion
		return fmt.Errorf("%w: no backup location set via --mysql-shell-backup-location", MySQLShellPreCheckError)
```

frouioui commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,510 @@
+		return fmt.Errorf("%w: no backup location set via --mysql_shell_location", MySQLShellPreCheckError)
+	}
+
+	if mysqlShellFlags == "" || !strings.Contains(mysqlShellFlags, "--js") {
+		return fmt.Errorf("%w: at least the --js flag is required", MySQLShellPreCheckError)
+	}
```
We should probably specify that "at least the --js flag is required" in the value of the flag `--mysql-shell-flags`. If we see this error in the logs it won't be clear if the vitess binary is missing the `--js` flag or if the value of the flag is missing content.

shlomi-noach commented:
Hey all! So we're all realizing this PR has been ongoing for a while. None of us was expecting it to take this long, and we keep finding new stuff we were unaware of at the beginning.

So first of all, please know we all appreciate your efforts and patience. We're all learning on the go I've personally never used `mysqlshell` before, let alone `mysqlshell` backups).

The other thing is I'm thinking there are multiple ongoing discussions on this issue. I wanted to propose we all take a brief pause, and figure out what the remaining issues are. Also, are y'all interested in synching to flesh out the remaining issues? Or would you like to prefer the discussion here?

rvrangel commented:
I think syncing would be great so we can sort any leftover issues

shlomi-noach commented:
Minutes of our sync meeting, attending @rvrangel, @mattlord , @Paoli, @shlomi-noach:

- The two main remaining issues are backup/restore of all schemas, and vitess-operator support
- We are going to document the MySQL shell engine support as "Experimental". As such, this PR needs not be fully compatible with every Vitess functionality. Specifically, we decide to defer vitess-operator support to a later stage (or work concurrently on https://github.com/planetscale/vitess-operator/pull/586).
- We have a better understanding of the Slack use case. I will not fully illustrate it here, but in general they only need to backup/restore a single database (plus `_vt`), and have their own credentials management automation.
- But that's a specialized use case. The general use case, which should be the default, is to backup and to restore the entire database. `mysqlshell` doesn't exactly support that, and the backup flow will be:
  1. Detect all non-system databases
  2. Backup each and every one of them
  3. Backup the MySQL users -- this is a different operation than a schema backup, and not consistent with the schema (ie not under the same `flush tables with read lock`)
  The restore flow will be:
  1. Delete existing schemas
  2. Delete users from `mysql.users`
  3. Load schemas from backup
  4. Load users
  5. Issue `RESET BINARY LOGS AND GTIDS` (formerly known as `RESET MASTER`), to remove any binary logs and to be able to set `@@global.gtid_executed`
  6. Set `@@global.gtid_executed`
  7. Resume replication
- The `backup_pitr_mysqlshell` CI test should do well to validate the full backup restore. After checking, this test framework does _not_ put the restored tablet into the replication stream -- we should do that, to validate we've populated `@@gtid_Executed` correctly. I can help with that. All that the test do is to apply backed up binary logs. The needed change is small and again I'm happy to work on that.
- We need documentation to clarify that changes to user accounts in between backup/restore may result with incorrect accounts/password on the restored server. The backup/restore process uses the `vt_dba` account, which generally does not ever change. So we will document this as a "known limitation" and we do have multiple potential paths to address this in the future. `vitess` does have the mechanism up update accounts/credentials.

Did I leave out anything?

rvrangel commented:
okay, so I have made the updates do remove the other databases and users a note on this:
- we ignore the same databases excluded by MySQL Shell by default
- for users, we ignore the reserved MySQL users defined in the docs
- it seems we can't create the `root` user after dropping it as the `vt_dba` (at least on the `endtoend` environment) doesn't have permissions to create all of the permissions (it was failing trying to `GRANT PROXY` on my tests), so I added it to the excluded list for now, let me know if this is a problem
- we talked about getting rid of the `DeleteBeforeRestore`, for now I will just ignore that flag (which is what the other engines are doing already) and it can be removed on a separate PR. I imagine doing this should unblock `vitess-operator`, so @frouioui I would appreciate if you could run the tests with the latest to see if that works

frouioui commented:
> * we talked about getting rid of the `DeleteBeforeRestore`, for now I will just ignore that flag (which is what the other engines are doing already) and it can be removed on a separate PR. I imagine doing this should unblock `vitess-operator`, so @frouioui I would appreciate if you could run the tests with the latest to see if that works

@rvrangel, awesome! I will to do that this week. Where is the best place to update you on the status, here or the vitess-operator PR?

rvrangel commented:
I would say you can update on the other PR since the plan is for us to merge this once the other things have been addressed, but up to you really :)

rvrangel commented:
@shlomi-noach I am checking with @frouioui if we can fix that issue with `vitess-operator` in this PR but other than that do you have anything else left to be addressed here other than the replication test? did you want that to be part of this PR too?

shlomi-noach commented:
@rvrangel - awesome, thank you -- yes, the replication test. As mentioned, I can help with that, if you like.

rvrangel commented:
yeah, that would be great, thanks :)

shlomi-noach commented:
All right. Give me some time and I'll be pushing into this branch!

shlomi-noach commented:
Merged and pushed https://github.com/vitessio/vitess/pull/16807 to be part of this PR, validating rejoining replication stream.

shlomi-noach commented:
Cross referencing the discussion around `root` account: https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2360486664, which we think is still important to solve here.

rvrangel commented:
sounds good, thanks for adding that test! I think the only way to fix this issue so that we can create the `root` user again would be to give all necessary permissions to `vt_dba` when it gets created. where do we do this today?

frouioui commented:
> sounds good, thanks for adding that test! I think the only way to fix this issue so that we can create the `root` user again would be to give all necessary permissions to `vt_dba` when it gets created. where do we do this today?

I think `init_db.sql` would be the right place, but correct me if i am wrong @shlomi-noach. We would also need to change the embedding of `init_db.sql` in the operator examples (like in `examples/operator/101_initial_cluster.yaml`). The test files on the vitess-operator repo will have have to be changed to include the new `GRANT`, this can be done in `test/endtoend/operator` of the vtop repo)

deepthi commented:
> we talked about getting rid of the DeleteBeforeRestore, for now I will just ignore that flag (which is what the other engines are doing already) and it can be removed on a separate PR

Can you please create a separate issue for this before doing a PR? `DeleteBeforeRestore` is set to true if someone calls the `RestoreFromBackup` RPC directly on a vttablet, but false if the tablet is being started with `--restore_from_backup` flag and potentially performing a restore. Removing it does not sound correct to me. That flag defaults to true, but we don't want to overwrite an existing non-empty database from a backup on vttablet restart.

rvrangel commented:
@deepthi Yes, I have created https://github.com/vitessio/vitess/issues/16811 now.

rvrangel commented:
@frouioui I have added the missing grant, hopefully that will fix it on the vitess operator. I have also removed `root` from the list of excluded users by default

frouioui commented:
@rvrangel, the latest changes look good to me, and the vtop tests are all passing!
However, we should add the grant in `examples/compose/config/init_db.sql` and `go/test/endtoend/vreplication/testdata/config/init_testserver_db.sql`

frouioui commented on .github/workflows/cluster_endtoend_backup_pitr_mysqlshell.yml:
Let's think about marking this new workflow as required when merging the PR.

shlomi-noach replied:
Agreed. We'll give it a couple weeks until all open PRs are in sync, then we'll make it `Required`.

frouioui commented on changelog/21.0/21.0.0/summary.md:
```diff
@@ -132,6 +133,13 @@ A query can also be set to have no timeout by using the `QUERY_TIMEOUT_MS` comme
 
+### <a id="new-backup-engine"/>New Backup Engine (EXPERIMENTAL)
+
+We are introducing a the backup engine supporting logical backups starting on v21 to support use cases that require something else besides physical backups. This is experimental and is based on the 
+[MySQL Shell](https://dev.mysql.com/doc/mysql-shell/8.0/en/).
```
```suggestion
We are introducing a backup engine supporting logical backups starting on v21 to support use cases that require something else besides physical backups. This is experimental and is based on the 
[MySQL Shell](https://dev.mysql.com/doc/mysql-shell/8.0/en/).
```

rvrangel replied:
thanks!

frouioui commented on changelog/21.0/21.0.0/summary.md:
```diff
@@ -132,6 +133,13 @@ A query can also be set to have no timeout by using the `QUERY_TIMEOUT_MS` comme
+We are introducing a the backup engine supporting logical backups starting on v21 to support use cases that require something else besides physical backups. This is experimental and is based on the 
+[MySQL Shell](https://dev.mysql.com/doc/mysql-shell/8.0/en/).
+
+The new engine is enabled by using `--backup_engine_implementation=mysqlshell`. There are other options that are required, so check the docs on which options are required and how to use it.
```
Adding a link to the docs would be useful here.

frouioui commented on go/vt/mysqlctl/backupengine.go:
```diff
@@ -179,6 +179,7 @@ func (p *RestoreParams) IsIncrementalRecovery() bool {
 // Returns the manifest of a backup if successful, otherwise returns an error
 type RestoreEngine interface {
 	ExecuteRestore(ctx context.Context, params RestoreParams, bh backupstorage.BackupHandle) (*BackupManifest, error)
+	ShouldStartMySQLAfterRestore() bool
```
👍🏻

frouioui commented on go/vt/mysqlctl/mysqlshellbackupengine.go:
```diff
@@ -0,0 +1,584 @@
+	return err
+}
+
+func init() {
+	BackupRestoreEngineMap[mysqlShellBackupEngineName] = &MySQLShellBackupEngine{}
+}
```
We already have an `init` at the top of this file, nit picking, but we could merge them together to make it easier to read.

rvrangel replied:
ah, good catch, I have merged them together

shlomi-noach reviewed (commented):
Approved on my side, pending final comments from @frouioui

shlomi-noach commented:
Merged and pushed `main` having merged https://github.com/vitessio/vitess/pull/16807, and to get latest fixes to unit test issues.

frouioui reviewed (commented):
Everything looks good to me besides https://github.com/vitessio/vitess/pull/16295#issuecomment-2364139208.

The vitess-operator PR also looks good to me and is ready to go.

rvrangel commented:
@frouioui thanks, I missed that comment. I have added the GRANT on those files now!

frouioui reviewed (approved):
Thank you for working with us through this big change! I don't think any of us anticipated the amount of work related to adding a new backup engine.

Vitess-operator, Docker, and the Vitess part look good to me!

rvrangel commented:
Thanks everyone for the support, it was a long PR but it definitely looks a much better solution than when it started! I appreciate all the suggestions feedback provided and great to see it finally merged 🎉
